File: Lightweight Generics Example/ModelController.m
//
//  DictionaryWrapper.m
//  Lightweight Generics Example
//
//  Created by Brian Gerstle on 10/1/15.
//  Copyright © 2015 Wikimedia Foundation. All rights reserved.
//

#import "ModelController.h"

@implementation ModelController

- (id)methodReturningIndex {
    return self.models.allKeys.firstObject;
}

- (id)methodReturningModel {
    return self.models.allValues.firstObject;
}

@end

@implementation Model

@end

@implementation SpecializedModelController

@end

File: Lightweight Generics Example/ModelController.swift
//
//  ModelController.swift
//  Lightweight Generics Example
//
//  Created by Brian Gerstle on 10/1/15.
//  Copyright © 2015 Wikimedia Foundation. All rights reserved.
//

import Foundation

func doSomethingWith(modelController: ModelController) {
    // generic methods
    let index: EmptyIndexProtocol! = modelController.methodReturningIndex()
    print(index)
    let model: EmptyModelProtocol! = modelController.methodReturningModel()
    print(model)

    // generic properties

    // sequence of NSObject
    let allKeys = modelController.models.keys
    print(allKeys)

    // sequence of EmptyModelProtocol
    let models = modelController.models.values
    print(models)
}

func doSomethingWith(modelController: SpecializedModelController) {
    // specialized generic methods
    let index: EmptyIndexProtocol! = modelController.methodReturningIndex()
    print(index)
    let model: EmptyModelProtocol! = modelController.methodReturningModel()
    print(model)

    // specialized generic properties

    // sequence of NSObject
    let allKeys = modelController.models.keys
    print(allKeys)

    // sequence of EmptyModelProtocol 
    let models = modelController.models.values
    print(models)
}
